### PR TITLE
lite2: migrate to proto

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

- migrate lite2 pkg to proto

Signed-off-by: Marko Baricevic <marbar3778@yahoo.com>

### CODE REVIEW

File: lite2/store/db/db.go [hunk 1]
```diff
@@ -56,12 +50,19 @@ func (s *dbs) SaveSignedHeaderAndValidatorSet(sh *types.SignedHeader, valSet *ty
 		panic("negative or zero height")
 	}
 
-	shBz, err := s.cdc.MarshalBinaryLengthPrefixed(sh)
+	pbsh := sh.ToProto()
+
+	shBz, err := proto.Marshal(pbsh)
+	if err != nil {
+		return errors.Wrap(err, "marshalling SignedHeader")
```
Comment (marbar3778) [inline lite2/store/db/db.go:57]: fmt.errorf

File: lite2/store/db/db_test.go [hunk 2]
```diff
@@ -26,7 +28,7 @@ func TestLast_FirstSignedHeaderHeight(t *testing.T) {
 
 	// 1 key
 	err = dbStore.SaveSignedHeaderAndValidatorSet(
-		&types.SignedHeader{Header: &types.Header{Height: 1}}, &types.ValidatorSet{})
+		&types.SignedHeader{Header: &types.Header{Height: 1, ProposerAddress: tmrand.Bytes(20)}}, vals)
```
Comment (melekes) [inline lite2/store/db/db_test.go:31]:
```suggestion
		&types.SignedHeader{Header: &types.Header{Height: 1, ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, vals)
```

File: lite2/store/db/db_test.go [hunk 3]
```diff
@@ -52,7 +54,7 @@ func Test_SaveSignedHeaderAndValidatorSet(t *testing.T) {
...
```
Comment (melekes) [inline lite2/store/db/db_test.go:57]:
```suggestion
		&types.SignedHeader{Header: &types.Header{Height: 1, ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, vals)
```

File: lite2/store/db/db_test.go [hunk 4]
```diff
@@ -78,14 +80,15 @@ func Test_SaveSignedHeaderAndValidatorSet(t *testing.T) {
...
+	valSet, _ := types.RandValidatorSet(10, 100)
 
 	assert.Panics(t, func() {
 		_, _ = dbStore.SignedHeaderBefore(0)
 		_, _ = dbStore.SignedHeaderBefore(100)
 	})
 
 	err := dbStore.SaveSignedHeaderAndValidatorSet(
-		&types.SignedHeader{Header: &types.Header{Height: 2}}, &types.ValidatorSet{})
+		&types.SignedHeader{Header: &types.Header{Height: 2, ProposerAddress: tmrand.Bytes(20)}}, valSet)
```
Comment (melekes) [inline lite2/store/db/db_test.go:91]:
```suggestion
		&types.SignedHeader{Header: &types.Header{Height: 2, ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, valSet)
```

File: lite2/store/db/db_test.go [hunk 5]
```diff
@@ -105,7 +109,7 @@ func Test_Prune(t *testing.T) {
 
 	// One header
 	err = dbStore.SaveSignedHeaderAndValidatorSet(
-		&types.SignedHeader{Header: &types.Header{Height: 2}}, &types.ValidatorSet{})
+		&types.SignedHeader{Header: &types.Header{Height: 2, ProposerAddress: tmrand.Bytes(20)}}, valSet)
```
Comment (melekes) [inline lite2/store/db/db_test.go:112]:
```suggestion
		&types.SignedHeader{Header: &types.Header{Height: 2, ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, valSet)
```

File: lite2/store/db/db_test.go [hunk 6]
```diff
@@ -121,7 +125,7 @@ func Test_Prune(t *testing.T) {
 	// Multiple headers
 	for i := 1; i <= 10; i++ {
 		err = dbStore.SaveSignedHeaderAndValidatorSet(
-			&types.SignedHeader{Header: &types.Header{Height: int64(i)}}, &types.ValidatorSet{})
+			&types.SignedHeader{Header: &types.Header{Height: int64(i), ProposerAddress: tmrand.Bytes(20)}}, valSet)
```
Comment (melekes) [inline lite2/store/db/db_test.go:128]:
```suggestion
			&types.SignedHeader{Header: &types.Header{Height: int64(i), ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, valSet)
```

File: lite2/store/db/db_test.go [hunk 7]
```diff
@@ -144,7 +149,7 @@ func Test_Concurrency(t *testing.T) {
 			defer wg.Done()
 
 			dbStore.SaveSignedHeaderAndValidatorSet(
-				&types.SignedHeader{Header: &types.Header{Height: i}}, &types.ValidatorSet{})
+				&types.SignedHeader{Header: &types.Header{Height: i, ProposerAddress: tmrand.Bytes(20)}}, vals)
```
Comment (melekes) [inline lite2/store/db/db_test.go:152]:
```suggestion
				&types.SignedHeader{Header: &types.Header{Height: i, ProposerAddress: tmrand.Bytes(crypto.AddressSize)}}, vals)
```

File: lite2/store/db/db_test.go (hunk 2 above)
Comment (melekes) [reply lite2/store/db/db_test.go:31]: why the need to fill ProposerAddress?

Comment (marbar3778) [reply lite2/store/db/db_test.go:31]: when we go back and forth from proto we call validatebasic which checks proposerAddr is not len(0)

Comment (melekes) [reply lite2/store/db/db_test.go]: what about other fields then? we aren't checking them?

Comment (marbar3778) [reply lite2/store/db/db_test.go]: we call ValidateHash in which we check if its > 0 && == tmhash.Size.that is why other fields pass

Comment (melekes) [reply lite2/store/db/db_test.go]: hmm... okay.. we should probably align these too. either we require all OR not. In the latter case, in validate we can add `len(proposeraddress) ==0 ||`

File: lite2/store/db/db_test.go [hunk 8]
```diff
@@ -144,7 +148,7 @@ func Test_Concurrency(t *testing.T) {
 			defer wg.Done()
 
 			dbStore.SaveSignedHeaderAndValidatorSet(
-				&types.SignedHeader{Header: &types.Header{Height: i}}, &types.ValidatorSet{})
+				&types.SignedHeader{Header: &types.Header{Height: i}}, vals)
```
Comment (melekes) [inline lite2/store/db/db_test.go:151]: why do we need vals here?

Comment (melekes) [reply lite2/store/db/db_test.go:151]: because proto will return an error?

Comment (marbar3778) [reply lite2/store/db/db_test.go:151]: on to proto it errors when it is a nil validator set.

**Pull request merged**